x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/categorization/CategorizeTextAggregator.java

```diff
@@ -0,0 +1,245 @@
+    protected LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub) throws IOException {
+        return new LeafBucketCollectorBase(sub, null) {
+            private final BytesRefBuilder scratch = new BytesRefBuilder();
+
+            @Override
+            public void collect(int doc, long owningBucketOrd) throws IOException {
+                collectFromSource(doc, owningBucketOrd);
+            }
+
+            private void collectFromSource(int doc, long owningBucketOrd) throws IOException {
+                sourceLookup.setSegmentAndDocument(ctx, doc);
+                Iterator<String> itr = sourceLookup.extractRawValues(sourceFieldName).stream().map(obj -> {
+                    if (obj == null) {
+                        return null;
+                    }
+                    if (obj instanceof BytesRef) {
+                        return fieldType.valueForDisplay(obj).toString();
+                    }
+                    return obj.toString();
+                }).iterator();
+                while (itr.hasNext()) {
+                    TokenStream ts = analyzer.tokenStream(fieldType.name(), itr.next());
+                    processTokenStream(owningBucketOrd, ts, doc);
+                }
+            }
+
+            private void processTokenStream(long owningBucketOrd, TokenStream ts, int doc) throws IOException {
+                ArrayList<Integer> tokens = new ArrayList<>();
+                try {
+                    CharTermAttribute termAtt = ts.addAttribute(CharTermAttribute.class);
+                    ts.reset();
+                    while (ts.incrementToken()) {
+                        tokens.add(bytesRefHash.put(new BytesRef(termAtt)));
+                    }
+                    if (tokens.isEmpty()) {
+                        return;
+                    }
+                } finally {
+                    ts.close();
+                }
+                categorizers = bigArrays().grow(categorizers, owningBucketOrd + 1);
+                CategorizationTokenTree categorizer = categorizers.get(owningBucketOrd);
+                if (categorizer == null) {
+                    categorizer = new CategorizationTokenTree(maxUniqueTokens, maxMatchTokens, similarityThreshold);
+                    addRequestCircuitBreakerBytes(categorizer.ramBytesUsed());
+                    categorizers.set(owningBucketOrd, categorizer);
+                }
+                long previousSize = categorizer.ramBytesUsed();
+                TextCategorization lg = categorizer.parseLogLine(
+                    tokens.stream().mapToInt(Integer::valueOf).toArray(),
+                    docCountProvider.getDocCount(doc)
+                );
+                long newSize = categorizer.ramBytesUsed();
+                if (newSize - previousSize > 0) {
+                    addRequestCircuitBreakerBytes(newSize - previousSize);
+                }
+
+                long bucketOrd = bucketOrds.add(owningBucketOrd, lg.getId());
```

Comment:
Oh boy. This kind of worries me because you can merge buckets as you go, right? If you immediately run sub-aggs now and you merge buckets later I think you'll get into trouble. We *do* have a way of merging document counts but we don't have a way of merging sub-aggregator buckets.

Reply:
Would you be ok not supporting sub-aggregations in a first pass?

Reply:
We have a thing, `DeferableBucketAggregator` that has support for these deferred children. The thing you are most similar to is `AutoDateHistogramAggregator`. But it'd add a bunch of complexity to do this right and might not be worth it on the first run. Maybe never?

Reply:
>  This kind of worries me because you can merge buckets as you go, right? 

This class does no merging of existing categories. We only expand existing ones.